### PR TITLE
[coin migration] deterministic address of paired FA metadata object

### PR DESCRIPTION
**Pull request by lightmark**
## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [x] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [x] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by davidiw — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/object.move [hunk 1]
```diff
@@ -348,7 +353,7 @@ module aptos_framework::object {
 
     /// Generates the DeriveRef, which can be used to create determnistic derived objects from the current object.
     public fun generate_derive_ref(ref: &ConstructorRef): DeriveRef {
-        DeriveRef { self: ref.self }
+        create_derive_ref(ref.self)
```
[the remark this answers is not in the document]
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/object.move:356]: kk

**Review by gregnazario — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/coin.move [hunk 2]
```diff
@@ -301,7 +301,10 @@ module aptos_framework::coin {
                 if (is_apt) {
                     object::create_sticky_object_at_address(@aptos_framework, @aptos_fungible_asset)
                 } else {
-                    object::create_sticky_object(coin_address<CoinType>())
+                    object::create_user_derived_object(
+                        @aptos_fungible_asset,
```
Comment (igor-aptos) [inline aptos-move/framework/aptos-framework/sources/coin.move:305]: why do we derive from @aptos_fungible_asset for non-apt ? that seems a bit awkward? is there nothing better to put here? even 0x1 sounds better

**Review by igor-aptos — CHANGES_REQUESTED** [review]

File: aptos-move/framework/aptos-framework/sources/coin.move (hunk 2 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move:305]: derive(owner_account_address, metadata_address) = (0xA, coin_creator_address)
So this is a trick that leverages this function to generate unique address w/o a new domain separator.
I think what you mean is `derive(coin_creator_address, 0xA)`?

Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move:305]: actually I was using (coin_creator_address, 0x0)... so not sure what's the best tuple here. cc @davidiw

Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/coin.move:305]: 0xA is APT FA address, and is unrelated to mapping here
`derive(coin_creator_address, 0x0) `seems better to me, or even `derive(coin_creator_address, coin_creator_address)`

Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move:305]: I like (coin_creator_address, coin_creator_address)... @davidiw sgty?

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/coin.move]: No, we cannot allow this to be influenced by a third party, someone could try to create that, no? I want this to be a guaranteed safe operation.

**Review by igor-aptos — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/coin.move (hunk 2 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move]: No, create_user_derived_object is a friend function that can only be called by 0x1.

Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move]: ` public(friend) fun create_user_derived_object(creator_address: address, derive_ref: &DeriveRef)`
it is a friend function so third party cannot create that

Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/coin.move]: so this function only exists for the purpose of primary fungible store address?

Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move]: Currently this is the only use case for now. But not necessarily be for PFS address in the future.

Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move]: also the first argument to this function is `address` so the value does not matter at all with regard to security.

File: aptos-move/framework/aptos-framework/sources/coin.move [hunk 3]
```diff
@@ -301,7 +301,11 @@ module aptos_framework::coin {
                 if (is_apt) {
                     object::create_sticky_object_at_address(@aptos_framework, @aptos_fungible_asset)
                 } else {
-                    object::create_sticky_object(coin_address<CoinType>())
+                    let coin_address = coin_address<CoinType>();
+                    object::create_sticky_object_at_address(
+                        @aptos_framework,
+                        object::create_user_derived_object_address(coin_address, coin_address)
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/coin.move:307]: What's with 2x coin_address here?

**Comment by lightmark** [comment]
@igor-aptos @davidiw we (@movekevin and I) realize `derive(address, address)` cannot give unique address for each `CoinType` as one coin_address can have multiple `CoinType` struct to be paired. So we have to turn to another solution:
```
create_named_object(@0x0/0xA, type_info::type_name<CoinType>())
```
instead to guarantee the uniqueness.

**Pull request merged**